File: Graph/101. Symmetric Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
bool isSym(TreeNode* L, TreeNode* R){
    if (L == NULL && R == NULL)
        return true;
    if (L != NULL && R != NULL && L->val == R->val)
        return isSym(L->left, R->right) && isSym(L->right, R->left);
    return false;
}
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return isSym(root, root);
    }
};

File: Graph/101. Symmetric Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        # iterative
        from queue import Queue
        q = Queue()
        q.put([root, root])
        while not q.empty():
            node1, node2 = q.get()
            if node1 == None and node2 == None:
                continue
            elif node1 == None or node2 == None:
                return 0
            elif node1.val != node2.val:
                return 0
            q.put([node1.left, node2.right])
            q.put([node1.right, node2.left])
        return 1        

        '''
        # recursive
        def isSym(left, right):
            if left == None and right == None:
                return 1
            if left != None and right != None and left.val == right.val:
                return isSym(left.left, right.right) and isSym(left.right, right.left)
            return 0
        return isSym(root, root)
        '''